### ISSUE TITLE
`IO#rewind` breaks read operations with custom encodings

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#11990]
`IO::Decoder` maintains an internal buffer of decoded bytes yet to be consumed (`@out_slice`), but this buffer is not reset whenever `IO#rewind` is called on its owner. This means the contents of the buffer could leak on subsequent reads:

```crystal
io = IO::Memory.new "ＡＢＣＤ".encode("UTF-32LE")
io.set_encoding "UTF-32LE"
io.read_char   # => 'Ａ'
io.rewind
io.read_char   # => 'Ｂ'
io.gets_to_end # => "ＣＤＡＢＣＤ"

io = IO::Memory.new "ＡＢＣＤ".encode("UTF-32LE")
io.set_encoding "UTF-32LE"
io.read_char   # => 'Ａ'
io.seek(8)
io.gets_to_end # => "ＢＣＤＣＤ"
```

A side effect is that `IO#pos` reports the byte position of the external buffered reads:

```crystal
io = IO::Memory.new "ＡＢＣＤ".encode("UTF-32LE")
io.set_encoding "UTF-32LE"
io.read_char
io.pos # => 16
io.rewind
io.read_char
io.pos # => 0

io = IO::Memory.new(("ＡＢＣＤ" * 1000).encode("UTF-32LE"))
io.set_encoding "UTF-32LE"
io.read_char
io.pos # => 4096
```

At this rate I believe custom encodings should be done through a new `IO` wrapper class that implements `IO::Buffered` with mandatory read buffering, rather than having an encoder-decoder pair in every `IO`.

**Comment by asterite** [comment]
I personally prefer how this is done in Ruby, where every IO has an encoding that you can set.

**Comment by straight-shoota** [comment]
I don't think this can be done similar to `IO::Buffered`. `IO::Buffered` is an implementation specialization that depends on the type of the IO backend. When it benefits from buffering, the author of an `IO` implementation can include `IO::Buffered` to get buffered behaviour.

Encodings on the other hand are typically already transparently applied and depend mostly on usage configuration, rather than being an inherent property of a specific IO backend.

**Comment by straight-shoota** [comment]
I like this approach ❤️ I've always wondered about the peculiar encoding API which seems odd to be present in just every `IO` instance. This is a good use case for composition. I might be missing some practical usability aspects from my image. But in terms of general software designm this looks like a good idea.

> The details about buffered reads / writes might be error-prone; nothing is written to b in the snippet above if the #sync= call is dropped.

I suppose `sync = true` could be the default for `IO::Encoded`?

> The new API has no special handling for the default UTF-8 behavior; an IO::Encoded.new(..., "UTF-8") will blindly use Iconv for what would be a near identity transformation.

But we can optimize that, right? It just needs to be implemented.

**Comment by asterite** [comment]
A bit more about why I like the Ruby way, despite it not being "good software design": https://dev.to/asterite/why-i-love-ruby-part-5-3caa

**Comment by HertzDevil** [comment]
The `#pos` issue might be solvable if we make it so that `IO::Decoder#read(io)` uses only `io.read_byte` to fill `@buffer`. We then repeatedly call `LibC.iconv` with increasing in-buffer sizes. For example, to decode `Bytes[0x03, 0x02, 0xF6, 0x01, 0x00]` from UTF-32LE, we need the following `iconv` calls:

* `Bytes[0x03]` -> `EINVAL` (calls `read_byte`)
* `Bytes[0x03, 0x02]` -> `EINVAL` (calls `read_byte`)
* `Bytes[0x03, 0x02, 0xF6]` -> `EINVAL` (calls `read_byte`)
* `Bytes[0x03, 0x02, 0xF6, 0x01]` -> `EILSEQ` (calls `read_byte`)
* If `invalid: :skip` is not set, raise `ArgumentError` here
* `Bytes[0x02]` -> `EINVAL`
* `Bytes[0x02, 0xF6]` -> `EINVAL`
* `Bytes[0x02, 0xF6, 0x01]` -> `EINVAL`
* `Bytes[0x02, 0xF6, 0x01, 0x00]` -> okay (calls `read_byte`)

This may have a rather significant performance impact, but perhaps it is unavoidable because one would have done the same anyway if the whole decoder is rewritten in pure Crystal.

It also might be tempting to use the current large-buffer implementation on `IO::Buffered`, but now `#unbuffered_pos` must be kept in sync with the original byte counts associated with the _decoded_ UTF-8 byte sequences in `IO::Decoder#@out_slice` as they are read. I don't think that will be easy.
